### PR TITLE
clear `OMDB_` environment variables when running omdb tests

### PR DESCRIPTION
**Pull request by davepacheco**
Prior to this, I happened to leave `OMDB_NEXUS_URL` set in the environment, and that would cause spurious test failures:

```
dap@ivanova omicron-merge $ OMDB_NEXUS_URL= cargo nextest run -p omicron-omdb
info: experimental features enabled: setup-scripts
   Compiling omicron-omdb v0.1.0 (/home/dap/omicron-merge/dev-tools/omdb)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 23.37s
    Starting 3 tests across 2 binaries (run ID: 6df449d3-d021-43c9-b3a1-1a4a0f3efcd9, nextest profile: default)
       SETUP [      1/1] crdb-seed: cargo run -p crdb-seed --profile test
             [ 00:00:00] [                                                                                                                                               ] 0/3:          Finished `test` profile [unoptimized + debuginfo] target(s) in 0.76s
     Running `target/debug/crdb-seed`
Aug 16 21:20:15.670 INFO Using existing CRDB seed tarball: `/dangerzone/omicron_tmp/crdb-base-dap/b7700187af915d7b8385d065ee657a04f893c52c4de7ff43476e2c7e47a43880.tar`
  SETUP PASS [      1/1] crdb-seed: cargo run -p crdb-seed --profile test
        FAIL [   3.250s] omicron-omdb::test_all_output test_omdb_usage_errors

--- STDOUT:              omicron-omdb::test_all_output test_omdb_usage_errors ---

running 1 test
@@ -437,61 +437,29 @@
 termination: Exited(2)
 ---------------------------------------------
 stdout:
 ---------------------------------------------
 stderr:
-Debug a specific Nexus instance
+error: 'omdb nexus' requires a subcommand but one was not provided
+  [subcommands: background-tasks, blueprints, sagas, sleds, help]
 
 Usage: omdb nexus [OPTIONS] <COMMAND>
 
-Commands:
-  background-tasks  print information about background tasks
-  blueprints        interact with blueprints
-  sagas             view sagas, create and complete demo sagas
-  sleds             interact with sleds
-  help              Print this message or the help of the given subcommand(s)
-
-Options:
-      --log-level <LOG_LEVEL>  log level filter [env: LOG_LEVEL=] [default: warn]
-  -h, --help                   Print help
-
-Connection Options:
-      --nexus-internal-url <NEXUS_INTERNAL_URL>  URL of the Nexus internal API [env:
-                                                 OMDB_NEXUS_URL=]
-      --dns-server <DNS_SERVER>                  [env: OMDB_DNS_SERVER=]
-
-Safety Options:
-  -w, --destructive  Allow potentially-destructive subcommands
+For more information, try '--help'.
 =============================================
 EXECUTING COMMAND: omdb ["nexus", "background-tasks"]
 termination: Exited(2)
 ---------------------------------------------
 stdout:
 ---------------------------------------------
 stderr:
-print information about background tasks
+error: 'omdb nexus background-tasks' requires a subcommand but one was not provided
+  [subcommands: doc, list, show, activate, help]
 
 Usage: omdb nexus background-tasks [OPTIONS] <COMMAND>
 
-Commands:
-  doc       Show documentation about background tasks
-  list      Print a summary of the status of all background tasks
-  show      Print human-readable summary of the status of each background task
-  activate  Activate one or more background tasks
-  help      Print this message or the help of the given subcommand(s)
-
-Options:
-      --log-level <LOG_LEVEL>  log level filter [env: LOG_LEVEL=] [default: warn]
-  -h, --help                   Print help
-
-Connection Options:
-      --nexus-internal-url <NEXUS_INTERNAL_URL>  URL of the Nexus internal API [env:
-                                                 OMDB_NEXUS_URL=]
-      --dns-server <DNS_SERVER>                  [env: OMDB_DNS_SERVER=]
-
-Safety Options:
-  -w, --destructive  Allow potentially-destructive subcommands
+For more information, try '--help'.
 =============================================
 EXECUTING COMMAND: omdb ["nexus", "background-tasks", "show", "--help"]
 termination: Exited(0)
 ---------------------------------------------
 stdout:
@@ -535,63 +503,29 @@
 termination: Exited(2)
 ---------------------------------------------
 stdout:
 ---------------------------------------------
 stderr:
-interact with blueprints
+error: 'omdb nexus blueprints' requires a subcommand but one was not provided
+  [subcommands: list, show, diff, delete, target, regenerate, import, help]
 
 Usage: omdb nexus blueprints [OPTIONS] <COMMAND>
 
-Commands:
-  list        List all blueprints
-  show        Show a blueprint
-  diff        Diff two blueprints
-  delete      Delete a blueprint
-  target      Interact with the current target blueprint
-  regenerate  Generate a new blueprint
-  import      Import a blueprint
-  help        Print this message or the help of the given subcommand(s)
-
-Options:
-      --log-level <LOG_LEVEL>  log level filter [env: LOG_LEVEL=] [default: warn]
-  -h, --help                   Print help
-
-Connection Options:
-      --nexus-internal-url <NEXUS_INTERNAL_URL>  URL of the Nexus internal API [env:
-                                                 OMDB_NEXUS_URL=]
-      --dns-server <DNS_SERVER>                  [env: OMDB_DNS_SERVER=]
-
-Safety Options:
-  -w, --destructive  Allow potentially-destructive subcommands
+For more information, try '--help'.
 =============================================
 EXECUTING COMMAND: omdb ["nexus", "sagas"]
 termination: Exited(2)
 ---------------------------------------------
 stdout:
 ---------------------------------------------
 stderr:
-view sagas, create and complete demo sagas
+error: 'omdb nexus sagas' requires a subcommand but one was not provided
+  [subcommands: list, demo-create, demo-complete, help]
 
 Usage: omdb nexus sagas [OPTIONS] <COMMAND>
 
-Commands:
-  list           List sagas run by this Nexus
-  demo-create    Create a "demo" saga
-  demo-complete  Complete a demo saga started with "demo-create"
-  help           Print this message or the help of the given subcommand(s)
-
-Options:
-      --log-level <LOG_LEVEL>  log level filter [env: LOG_LEVEL=] [default: warn]
-  -h, --help                   Print help
-
-Connection Options:
-      --nexus-internal-url <NEXUS_INTERNAL_URL>  URL of the Nexus internal API [env:
-                                                 OMDB_NEXUS_URL=]
-      --dns-server <DNS_SERVER>                  [env: OMDB_DNS_SERVER=]
-
-Safety Options:
-  -w, --destructive  Allow potentially-destructive subcommands
+For more information, try '--help'.
 =============================================
 EXECUTING COMMAND: omdb ["nexus", "--nexus-internal-url", "http://[::1]:111", "sagas", "demo-create"]
 termination: Exited(1)
 ---------------------------------------------
 stdout:
@@ -604,32 +538,16 @@
 termination: Exited(2)
 ---------------------------------------------
 stdout:
 ---------------------------------------------
 stderr:
-interact with sleds
+error: 'omdb nexus sleds' requires a subcommand but one was not provided
+  [subcommands: list-uninitialized, add, expunge, expunge-disk, help]
 
 Usage: omdb nexus sleds [OPTIONS] <COMMAND>
 
-Commands:
-  list-uninitialized  List all uninitialized sleds
-  add                 Add an uninitialized sled
-  expunge             Expunge a sled (DANGEROUS)
-  expunge-disk        Expunge a disk (DANGEROUS)
-  help                Print this message or the help of the given subcommand(s)
-
-Options:
-      --log-level <LOG_LEVEL>  log level filter [env: LOG_LEVEL=] [default: warn]
-  -h, --help                   Print help
-
-Connection Options:
-      --nexus-internal-url <NEXUS_INTERNAL_URL>  URL of the Nexus internal API [env:
-                                                 OMDB_NEXUS_URL=]
-      --dns-server <DNS_SERVER>                  [env: OMDB_DNS_SERVER=]
-
-Safety Options:
-  -w, --destructive  Allow potentially-destructive subcommands
+For more information, try '--help'.
 =============================================
 EXECUTING COMMAND: omdb ["sled-agent"]
 termination: Exited(2)
 ---------------------------------------------
 stdout:

test test_omdb_usage_errors ... FAILED

failures:

failures:
    test_omdb_usage_errors

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 2 filtered out; finished in 3.22s


--- STDERR:              omicron-omdb::test_all_output test_omdb_usage_errors ---
thread 'test_omdb_usage_errors' panicked at dev-tools/omdb/tests/test_all_output.rs:110:5:
assertion failed: string doesn't match the contents of file: "tests/usage_errors.out" see diffset above
                set EXPECTORATE=overwrite if these changes are intentional
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

   Canceling due to test failure: 2 tests still running
```

In these cases, clap produced different help output because of the presence of the environment variable (see clap-rs/clap#5673).  This PR causes the omdb tests to clear all OMDB-related environment variables before doing anything else.  This is a good idea anyway since these affect many other omdb behaviors that are being tested.

### CODE REVIEW

File: nexus/test-utils/src/lib.rs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 //! Integration testing facilities for Nexus
 
+#[cfg(feature = "omicron-dev")]
```
Comment (davepacheco) [inline nexus/test-utils/src/lib.rs:7]: I threw this in because I saw a spurious warning about this import when doing `cargo nextest run -p omicron-omdb`.

File: dev-tools/omdb/tests/test_all_output.rs [hunk 2]
```diff
@@ -504,3 +509,19 @@ async fn do_run_extra<F>(
...
+
+// OMDB environment variables can affect even the help output provided by clap.
+// See clap-rs/clap#5673 for an example.  This can cause spurious test failures.
+// Clear all the OMDB-related variables from the current process so that all
+// child processes don't have them.
+fn clear_omdb_env() {
+    // Rust documents that this is not safe to do in a multi-threaded process
+    // outside of Windows.  We cannot be sure that we are single-threaded,
+    // and we won't be if we're run with `cargo test`.  But we will be
+    // single-threaded in `cargo nextest` (the officially supported runner).
```
Comment (sunshowers) [inline dev-tools/omdb/tests/test_all_output.rs:521]: So "single-threaded" isn't quite accurate here... libtest will spin up a thread for the test on top of the main thread. But this is completely fine here because the main thread will just be blocked waiting for the test thread.

The important fact here is just that with nextest, we know that nothing else is altering the environment at the same time. So I'd maybe rephrase this in terms of "process-per-test".

**Comment by davepacheco** [comment]
For completeness, this is what it looks like now:

```
dap@ivanova omicron-merge $ OMDB_NEXUS_URL= cargo nextest run  -p omicron-omdb 
info: experimental features enabled: setup-scripts
    Finished `test` profile [unoptimized + debuginfo] target(s) in 1.01s
    Starting 3 tests across 2 binaries (run ID: 4313c614-e5e4-4fdf-a85d-49ac1b2b5702, nextest profile: default)
       SETUP [      1/1] crdb-seed: cargo run -p crdb-seed --profile test
             [ 00:00:00] [                                                                                                                                               ] 0/3:          Finished `test` profile [unoptimized + debuginfo] target(s) in 0.80s
     Running `target/debug/crdb-seed`
Aug 16 21:32:12.147 INFO Using existing CRDB seed tarball: `/dangerzone/omicron_tmp/crdb-base-dap/b7700187af915d7b8385d065ee657a04f893c52c4de7ff43476e2c7e47a43880.tar`
  SETUP PASS [      1/1] crdb-seed: cargo run -p crdb-seed --profile test
        PASS [   3.208s] omicron-omdb::test_all_output test_omdb_usage_errors
        PASS [  11.889s] omicron-omdb::test_all_output test_omdb_env_settings
        PASS [  16.134s] omicron-omdb::test_all_output test_omdb_success_cases
------------
     Summary [  17.091s] 3 tests run: 3 passed, 0 skipped
dap@ivanova omicron-merge $ 


dap@ivanova omicron-merge $ OMDB_NEXUS_URL= cargo nextest run  -p omicron-omdb --nocapture test_omdb_usage_errors
info: experimental features enabled: setup-scripts
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.99s
    Starting 1 test across 2 binaries (2 skipped; run ID: 092613e1-5df8-4f05-936a-4b0d358317c9, nextest profile: default)
       SETUP [      1/1] crdb-seed: cargo run -p crdb-seed --profile test
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.81s
     Running `target/debug/crdb-seed`
Aug 16 21:31:38.004 INFO Using existing CRDB seed tarball: `/dangerzone/omicron_tmp/crdb-base-dap/b7700187af915d7b8385d065ee657a04f893c52c4de7ff43476e2c7e47a43880.tar`
  SETUP PASS [      1/1] crdb-seed: cargo run -p crdb-seed --profile test
       START             omicron-omdb::test_all_output test_omdb_usage_errors

running 1 test
removing "OMDB_NEXUS_URL" from environment
test test_omdb_usage_errors ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 2 filtered out; finished in 2.87s

        PASS [   2.900s] omicron-omdb::test_all_output test_omdb_usage_errors
------------
     Summary [   3.853s] 1 test run: 1 passed, 2 skipped
```

**Review by sunshowers — APPROVED** [review]
Thanks!

File: dev-tools/omdb/tests/test_all_output.rs (hunk 2 above)
Comment (sunshowers) [reply dev-tools/omdb/tests/test_all_output.rs:521]: And there's also separate safety and correctness issues here.

* Two threads using the Rust `std::env::remove_var` to alter the environment is safe but may not be correct.
* One thread using `std::env::remove_var` is both safe and probably correct.
* One thread using `std::env::remove_var` and another thread using libc's `unsetenv` is unsafe on many Unix platforms (though it sounds like not illumos).

Comment (davepacheco) [reply dev-tools/omdb/tests/test_all_output.rs]: I edited this comment (and the one before it, since it was implying that the clap issue was the only one here, which isn't true)
